lib/tasks/notifications.rake

```diff
@@ -1,21 +1,35 @@
 namespace :notifications do
   desc 'Send webhook and email notifications to the specified supplier for future moves'
-  task :send, %i[supplier from_date send_webhooks send_emails] => :environment do |_, args|
+  task :send, %i[supplier from_date send_webhooks send_emails resend action_name] => :environment do |_, args|
     abort "Please specify a supplier, e.g. $ rake 'notifications:test_webhook[serco]'" if args[:supplier].blank?
     supplier = Supplier.find_by(key: args[:supplier]) || Supplier.find_by(id: args[:supplier]) || Supplier.find_by(name: args[:supplier])
     abort "Unknown supplier: #{args[:supplier]}" if supplier.blank?
 
     from_date = args[:from_date].present? ? Date.parse(args[:from_date]) : Time.zone.today
     send_webhooks = args[:send_webhooks].present? ? args[:send_webhooks] == 'true' : true # default to sending webhooks
     send_emails = args[:send_emails].present? ? args[:send_emails] == 'true' : false # default to not sending emails
+    resend = args[:resend].present? ? args[:resend] == 'true' : false # default to not re-sending notifications if they have already been sent
+    action_name = args[:action_name] || 'create'
 
     statuses = [Move::MOVE_STATUS_REQUESTED, Move::MOVE_STATUS_BOOKED]
 
+    notification_types = []
+    notification_types << NotificationType::WEBHOOK if send_webhooks
+    notification_types << NotificationType::EMAIL if send_emails
+
+    already_sent_notifications_sql = Notification
+                                         .where.not(delivered_at: nil)
+                                         .where(topic_type: 'Move')
+                                         .where(notification_type: notification_types)
+                                         .select(:topic_id)
+                                         .to_sql
+
     # moves directly assigned to the supplier
     moves_assigned = Move
       .where(supplier: supplier)
       .where('date >= ?', from_date)
       .where(status: statuses)
+      .where(resend ? '' : "id NOT IN (#{already_sent_notifications_sql})")
```

Comment:
Not seen this technique used before `where('')` - surprised it works actually. Just a minor thing but maybe worth splitting out separately for readability. Also maybe it's simpler to just fetch a list of the move id's that have already been processed and pass that as a parameter to a more idiomatic Active Record chain. Something like this...

```ruby
if resend
  previously_sent_move_ids = Notification
                                         .where.not(delivered_at: nil)
                                         .where(topic_type: 'Move')
                                         .where(notification_type: notification_types)
                                         .select(:topic_id)
  moves_assigned = moves_assigned.where.not(id: previously_sent_move_ids)
end
```

Reply:
so, I don't want to pass the `id`s directly to active record as there could be tens of thousands - the sql approach, although a bit clunky is definitely more efficient. However, I've re-written it along the lines you suggest to improve readability.

BTW the ActiveRecord trick:  `where(nil)`, `where({})` and `where("")` works and can be handy sometimes for conditional querying.

Reply:
Fair enough, better to avoid passing a huge list of ID's to Postgres and good to learn something new 😄